src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs

```diff
@@ -17,34 +17,98 @@ namespace System.Buffers
+        {
+            first = default;
+            next = default;
+            SequencePosition start = buffer.Start;
+            int startIndex = start.GetInteger();
+            object startObject = start.GetObject();
+
+            if (startObject != null)
+            {
+                SequencePosition end = buffer.End;
+                int endIndex = end.GetInteger();
+                bool isMultiSegment = startObject != end.GetObject();
+
+                // A == 0 && B == 0 means SequenceType.MultiSegment
+                if (startIndex >= 0)
                 {
-                    CurrentSpan = default;
-                    // No space in the first span, move to one with space
-                    GetNextSpan();
+                    if (endIndex >= 0)  // SequenceType.MultiSegment
+                    {
+                        ReadOnlySequenceSegment<T> segment = (ReadOnlySequenceSegment<T>)startObject;
+                        next = new SequencePosition(segment.Next, 0);
+                        first = segment.Memory.Span;
+                        if (isMultiSegment)
+                        {
+                            first = first.Slice(startIndex);
+                        }
+                        else
+                        {
+                            first = first.Slice(startIndex, endIndex - startIndex);
+                        }
+                    }
+                    else
+                    {
+                        if (isMultiSegment)
+                            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+
+                        first = new ReadOnlySpan<T>((T[])startObject, startIndex, (endIndex & IndexBitMask) - startIndex);
+                    }
                 }
                 else
                 {
-                    CurrentSpan = memory.Span;
+                    first = GetFirstSpanSlow(startObject, startIndex, endIndex, isMultiSegment);
                 }
             }
-            else
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static ReadOnlySpan<T> GetFirstSpanSlow(object startObject, int startIndex, int endIndex, bool isMultiSegment)
+        {
+            Debug.Assert(startIndex < 0 || endIndex < 0);
+            if (isMultiSegment)
+                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+
+            // The type == char check here is redundant. However, we still have it to allow
+            // the JIT to see when that the code is unreachable and eliminate it.
+            // A == 1 && B == 1 means SequenceType.String
+            if (typeof(T) == typeof(char) && endIndex < 0)
             {
-                // No space in any spans and at end of sequence
-                _moreData = false;
-                CurrentSpan = default;
+                var memory = (ReadOnlyMemory<T>)(object)((string)startObject).AsMemory();
```

Comment:
Note that once this moves to ReadOnlySequence, it can be re-written as follows:
```C#
ReadOnlySpan<char> spanOfChar =((string)startObject).AsSpan(startIndex & IndexBitMask, endIndex - startIndex);
ReadOnlySpan<T> spanOfT = MemoryMarshal.CreateSpan(ref Unsafe.As<char, T>(ref MemoryMarshal.GetReference(spanOfChar)), spanOfChar.Length);
```

`MemoryMarshal.CreateSpan` is a netcoreapp only API.

Reply:
Noted